qiskit/quantum_info/quaternion.py

```diff
@@ -43,7 +43,7 @@ def __mul__(self, r):
             out_data[3] = r(0) * q(3) - r(1) * q(2) + r(2) * q(1) + r(3) * q(0)
             return Quaternion(out_data)
         else:
-            raise Exception("Multiplication by other not supported.")
+            raise RuntimeError("Multiplication by other not supported.")
```

Comment:
This should probably `return NotImplemented` rather than manually raise an exception.